File: examples/streaming_daemon.php
#!/usr/bin/env php
<?php
/**
 * Streaming Daemon for mod_socket_handoff (PHP version)
 *
 * Receives client connections from Apache via SCM_RIGHTS file descriptor
 * passing and streams responses directly to clients.
 *
 * Requirements:
 *   - PHP 7.4+ with sockets extension
 *   - pcntl extension (for signal handling)
 *
 * Run:
 *   sudo php streaming_daemon.php
 *
 * Note: This is a single-process daemon. For production, consider using
 * pcntl_fork() for each connection or a process manager like supervisord.
 */

declare(strict_types=1);

const SOCKET_PATH = '/var/run/streaming-daemon.sock';

/**
 * Receive a file descriptor and data via SCM_RIGHTS
 *
 * Uses socket_cmsg_space() to properly allocate the control message buffer.
 */
function receive_fd(Socket $socket): array
{
    $message = [
        'name' => [],
        'buffer_size' => 4096,
        'controllen' => socket_cmsg_space(SOL_SOCKET, SCM_RIGHTS, 1),
    ];

    $bytes = socket_recvmsg($socket, $message, 0);
    if ($bytes === false) {
        throw new RuntimeException('socket_recvmsg failed: ' . socket_strerror(socket_last_error($socket)));
    }

    // Extract the file descriptor from control message
    $fd = null;
    if (!empty($message['control'])) {
        foreach ($message['control'] as $cmsg) {
            if ($cmsg['level'] === SOL_SOCKET && $cmsg['type'] === SCM_RIGHTS) {
                // The data contains a resource/socket
                $fd = $cmsg['data'][0];
                break;
            }
        }
    }

    // Extract the data from iov
    $data = '';
    if (!empty($message['iov']) && !empty($message['iov'][0])) {
        $data = $message['iov'][0];
    }

    return [$fd, $data];
}

/**
 * Stream a simulated LLM response to the client
 *
 * @param resource|Socket|int $client_fd The client socket (resource, Socket object, or fd int)
 */
function stream_response($client_fd, array $handoff_data): void
{
    // Handle different types of client_fd
    if ($client_fd instanceof Socket) {
        // Convert Socket to stream
        $stream = socket_export_stream($client_fd);
        if ($stream === false) {
            error_log("Failed to export Socket to stream");
            return;
        }
    } elseif (is_resource($client_fd)) {
        // Already a resource/stream
        $stream = $client_fd;
    } elseif (is_int($client_fd)) {
        // File descriptor integer
        $stream = fopen("php://fd/{$client_fd}", 'w');
        if ($stream === false) {
            error_log("Failed to open fd {$client_fd} as stream");
            return;
        }
    } else {
        error_log("Unknown client_fd type: " . gettype($client_fd));
        return;
    }

    // Disable buffering
    stream_set_write_buffer($stream, 0);

    // Send HTTP response headers
    $headers = "HTTP/1.1 200 OK\r\n";
    $headers .= "Content-Type: text/event-stream\r\n";
    $headers .= "Cache-Control: no-cache\r\n";
    $headers .= "Connection: close\r\n";
    $headers .= "X-Accel-Buffering: no\r\n";
    $headers .= "\r\n";
    fwrite($stream, $headers);
    fflush($stream);

    // Prepare the response content
    $user_id = $handoff_data['user_id'] ?? 0;
    $prompt = $handoff_data['prompt'] ?? 'unknown';
    $prompt_display = strlen($prompt) > 100 ? substr($prompt, 0, 100) . '...' : $prompt;

    $lines = [
        "Hello! I received your request.",
        "",
        "User ID: {$user_id}",
        "Prompt: \"{$prompt_display}\"",
        "",
        "Let me help you with that. Here's my response:",
        "",
        "The Apache worker that handled your initial request has already been freed. " .
            "This response is being streamed directly from a lightweight PHP daemon " .
            "that received the client socket via SCM_RIGHTS file descriptor passing.",
        "",
        "This architecture allows:",
        "1. Heavy PHP processes to handle authentication",
        "2. Lightweight daemons to handle long-running streams",
        "3. Better resource utilization under load",
        "",
        "The connection handoff is transparent to the client - " .
            "you see a single HTTP request with a streaming response.",
        "",
        "This is the end of my response. Have a great day!",
    ];

    // Stream each line with simulated LLM delays
    foreach ($lines as $index => $line) {
        $data = json_encode(['content' => $line, 'index' => $index]);
        fwrite($stream, "data: {$data}\n\n");
        fflush($stream);

        // Simulate LLM thinking time - varies by line length
        $delay = 100000 + (strlen($line) * 10000); // microseconds
        if ($delay > 500000) {
            $delay = 500000;
        }
        usleep($delay);
    }

    // Send done marker
    fwrite($stream, "data: [DONE]\n\n");
    fflush($stream);

    fclose($stream);
}

/**
 * Handle a connection from Apache
 */
function handle_connection(Socket $conn): void
{
    try {
        [$client_fd, $data] = receive_fd($conn);

        if ($client_fd === null) {
            error_log("No file descriptor received");
            return;
        }

        $fd_type = is_resource($client_fd) ? 'resource' : (($client_fd instanceof Socket) ? 'Socket' : gettype($client_fd));
        echo "Received handoff: fd_type={$fd_type}, data_len=" . strlen($data) . "\n";

        // Parse handoff data
        $handoff_data = [];
        if (!empty($data)) {
            $handoff_data = json_decode($data, true) ?? [];
        }

        $user_id = $handoff_data['user_id'] ?? 'unknown';
        $prompt = $handoff_data['prompt'] ?? 'unknown';
        $prompt_short = strlen($prompt) > 50 ? substr($prompt, 0, 50) . '...' : $prompt;
        echo "  user_id={$user_id}, prompt=\"{$prompt_short}\"\n";

        // Stream response to client
        stream_response($client_fd, $handoff_data);

        // Close the client socket/resource
        if ($client_fd instanceof Socket) {
            socket_close($client_fd);
        } elseif (is_resource($client_fd)) {
            fclose($client_fd);
        }

        echo "  Connection closed\n";

    } catch (Throwable $e) {
        error_log("Error handling connection: " . $e->getMessage());
    }
}

/**
 * Main daemon loop
 */
function main(): void
{
    // Check for required extensions
    if (!extension_loaded('sockets')) {
        die("Error: sockets extension is required\n");
    }

    // Remove stale socket
    if (file_exists(SOCKET_PATH)) {
        unlink(SOCKET_PATH);
    }

    // Create Unix socket
    $server = socket_create(AF_UNIX, SOCK_STREAM, 0);
    if ($server === false) {
        die("socket_create failed: " . socket_strerror(socket_last_error()) . "\n");
    }

    // Bind to socket path
    if (!socket_bind($server, SOCKET_PATH)) {
        die("socket_bind failed: " . socket_strerror(socket_last_error($server)) . "\n");
    }

    // Set permissions so Apache can connect
    chmod(SOCKET_PATH, 0666);

    // Listen for connections
    if (!socket_listen($server, 5)) {
        die("socket_listen failed: " . socket_strerror(socket_last_error($server)) . "\n");
    }

    echo "Streaming daemon (PHP) listening on " . SOCKET_PATH . "\n";

    // Set up signal handling for graceful shutdown
    if (extension_loaded('pcntl')) {
        pcntl_signal(SIGINT, function () use ($server) {
            echo "\nShutting down...\n";
            socket_close($server);
            if (file_exists(SOCKET_PATH)) {
                unlink(SOCKET_PATH);
            }
            exit(0);
        });
        pcntl_signal(SIGTERM, function () use ($server) {
            echo "\nShutting down...\n";
            socket_close($server);
            if (file_exists(SOCKET_PATH)) {
                unlink(SOCKET_PATH);
            }
            exit(0);
        });
    }

    // Accept connections
    while (true) {
        // Check for signals if pcntl is available
        if (extension_loaded('pcntl')) {
            pcntl_signal_dispatch();
        }

        // Accept with timeout so we can check signals
        $read = [$server];
        $write = null;
        $except = null;
        $changed = socket_select($read, $write, $except, 1);

        if ($changed === false) {
            error_log("socket_select failed: " . socket_strerror(socket_last_error()));
            continue;
        }

        if ($changed === 0) {
            continue; // Timeout, check signals and loop
        }

        $conn = socket_accept($server);
        if ($conn === false) {
            error_log("socket_accept failed: " . socket_strerror(socket_last_error($server)));
            continue;
        }

        handle_connection($conn);
        socket_close($conn);
    }
}

// Run the daemon
main();

File: examples/streaming_daemon.go
/*
 * Streaming Daemon Example for mod_socket_handoff
 *
 * This daemon receives client connections from Apache via SCM_RIGHTS
 * and streams responses directly to clients. Use this as a template
 * for integrating with LLM APIs or other streaming services.
 *
 * Build:
 *   go build -o streaming-daemon streaming_daemon.go
 *
 * Run:
 *   sudo ./streaming-daemon
 *
 * Or install as systemd service - see streaming-daemon.service
 */

package main

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"runtime/debug"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"
)

const (
	// Socket path - must match SocketHandoffAllowedPrefix in Apache config
	DaemonSocket = "/var/run/streaming-daemon.sock"

	// Timeouts for robustness
	HandoffTimeout  = 5 * time.Second   // Max time to receive fd from Apache
	WriteTimeout    = 30 * time.Second  // Max time for a single write to client
	ShutdownTimeout = 2 * time.Minute   // Graceful shutdown timeout; long enough for LLM streams to complete
	MaxConnections  = 1000              // Max concurrent connections

	// MaxHandoffDataSize is the buffer size for receiving handoff JSON from Apache
	// via the Unix socket. The data originates from the X-Handoff-Data HTTP header.
	// This should be large enough to hold LLM prompts. Increase if needed.
	// Note: Apache's LimitRequestFieldSize (default 8190) limits individual header size.
	MaxHandoffDataSize = 65536 // 64KB
)

// HandoffData is the JSON structure passed from PHP via X-Handoff-Data header.
// Customize this struct to match your application's needs.
type HandoffData struct {
	UserID    int64  `json:"user_id"`
	Prompt    string `json:"prompt"`
	Model     string `json:"model,omitempty"`
	MaxTokens int    `json:"max_tokens,omitempty"`
	Timestamp int64  `json:"timestamp,omitempty"`
}

// Connection tracking for graceful shutdown
var (
	activeConns   int64
	connSemaphore chan struct{}
	connWg        sync.WaitGroup
)

func main() {
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	// Initialize connection limiter
	connSemaphore = make(chan struct{}, MaxConnections)

	// Handle shutdown gracefully
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)

	go func() {
		for sig := range sigChan {
			switch sig {
			case syscall.SIGHUP:
				log.Printf("Received SIGHUP, active connections: %d", atomic.LoadInt64(&activeConns))
			default:
				log.Printf("Received %v, shutting down...", sig)
				signal.Stop(sigChan)
				cancel()
				return
			}
		}
	}()

	// Remove stale socket
	os.Remove(DaemonSocket)

	// Create Unix socket listener
	listener, err := net.Listen("unix", DaemonSocket)
	if err != nil {
		log.Fatalf("Failed to listen on %s: %v", DaemonSocket, err)
	}
	// Ensure listener is closed on exit; also closed explicitly during graceful shutdown
	defer listener.Close()

	// Set permissions so Apache (www-data) can connect.
	// SECURITY: 0666 allows any local user to connect. For production,
	// consider 0660 with proper group ownership, or use directory permissions.
	if err := os.Chmod(DaemonSocket, 0666); err != nil {
		log.Printf("Warning: Could not chmod socket: %v", err)
	}

	log.Printf("Streaming daemon listening on %s (max %d connections)", DaemonSocket, MaxConnections)

	// Accept connections
	go func() {
		for {
			conn, err := listener.Accept()
			if err != nil {
				select {
				case <-ctx.Done():
					return
				default:
					log.Printf("Accept error: %v", err)
					time.Sleep(100 * time.Millisecond) // Back off on errors
					continue
				}
			}

			// Check if context is cancelled before processing connection
			select {
			case <-ctx.Done():
				if err := conn.Close(); err != nil {
					log.Printf("Error closing connection during shutdown: %v", err)
				}
				return
			default:
			}

			// Acquire semaphore (limit concurrent connections)
			select {
			case connSemaphore <- struct{}{}:
				connWg.Add(1)
				atomic.AddInt64(&activeConns, 1)
				go func(c net.Conn) {
					defer func() {
						<-connSemaphore
						connWg.Done()
						atomic.AddInt64(&activeConns, -1)
					}()
					// Check context again inside goroutine to handle race condition
					select {
					case <-ctx.Done():
						if err := c.Close(); err != nil {
							log.Printf("Error closing connection during shutdown: %v", err)
						}
						return
					default:
					}
					safeHandleConnection(ctx, c)
				}(conn)
			default:
				log.Printf("Connection limit reached (%d), rejecting", MaxConnections)
				if err := conn.Close(); err != nil {
					log.Printf("Error closing rejected connection: %v", err)
				}
				// Back off slightly to avoid tight accept-reject loops under overload,
				// but exit promptly if context is cancelled.
				select {
				case <-ctx.Done():
					return
				case <-time.After(10 * time.Millisecond):
				}
			}
		}
	}()

	<-ctx.Done()

	// Listener will be closed by defer; wait for active connections to finish
	log.Printf("Waiting for %d active connections to finish...", atomic.LoadInt64(&activeConns))
	done := make(chan struct{})
	go func() {
		connWg.Wait()
		close(done)
	}()

	select {
	case <-done:
		log.Println("All connections closed gracefully")
	case <-time.After(ShutdownTimeout):
		log.Printf("Timeout waiting for connections, %d still active", atomic.LoadInt64(&activeConns))
	}

	// Cleanup
	os.Remove(DaemonSocket)
	log.Println("Daemon stopped")
}

// safeHandleConnection wraps handleConnection with panic recovery
func safeHandleConnection(ctx context.Context, conn net.Conn) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("Panic in connection handler: %v\n%s", r, debug.Stack())
		}
	}()
	handleConnection(ctx, conn)
}

func handleConnection(ctx context.Context, conn net.Conn) {
	defer func() {
		if err := conn.Close(); err != nil {
			log.Printf("Error closing connection: %v", err)
		}
	}()

	// Receive file descriptor and handoff data from Apache
	clientFd, data, err := receiveFd(conn)
	if err != nil {
		log.Printf("Failed to receive fd: %v", err)
		return
	}

	// Wrap fd in os.File immediately to ensure cleanup on any error path.
	// os.NewFile takes ownership only on success; if it returns nil (invalid fd),
	// we must close the fd ourselves with syscall.Close().
	clientFile := os.NewFile(uintptr(clientFd), "client")
	if clientFile == nil {
		log.Printf("Failed to create file from fd %d", clientFd)
		syscall.Close(clientFd)
		return
	}
	defer clientFile.Close()

	log.Printf("Received handoff: fd=%d, data_len=%d", clientFd, len(data))

	// Parse handoff data
	var handoff HandoffData
	if len(data) > 0 {
		if err := json.Unmarshal(data, &handoff); err != nil {
			log.Printf("Failed to parse handoff data: %v", err)
		}
	}

	log.Printf("  user_id=%d, prompt=%q", handoff.UserID, truncate(handoff.Prompt, 50))

	// Stream response to client
	if err := streamToClient(ctx, clientFile, handoff); err != nil {
		log.Printf("Stream error: %v", err)
	}

	log.Printf("  Connection closed")
}

// receiveFd receives a file descriptor and data from the Unix socket connection.
// Apache sends the client socket fd via SCM_RIGHTS along with the handoff data.
func receiveFd(conn net.Conn) (int, []byte, error) {
	unixConn, ok := conn.(*net.UnixConn)
	if !ok {
		return -1, nil, fmt.Errorf("not a Unix connection")
	}

	file, err := unixConn.File()
	if err != nil {
		return -1, nil, fmt.Errorf("could not get file: %w", err)
	}
	defer file.Close()

	// Set receive timeout on the raw fd using SO_RCVTIMEO (Unix-specific).
	// We use SO_RCVTIMEO instead of conn.SetReadDeadline because we call
	// syscall.Recvmsg directly on the fd, bypassing Go's net.Conn deadline logic.
	// Use NsecToTimeval for portability across Unix platforms (Timeval field types vary).
	tv := syscall.NsecToTimeval(HandoffTimeout.Nanoseconds())
	if err := syscall.SetsockoptTimeval(int(file.Fd()), syscall.SOL_SOCKET, syscall.SO_RCVTIMEO, &tv); err != nil {
		return -1, nil, fmt.Errorf("could not set receive timeout: %w", err)
	}

	// Buffer for handoff data (sized to hold large LLM prompts)
	buf := make([]byte, MaxHandoffDataSize)

	// Buffer for control message (SCM_RIGHTS)
	// Size: 16 bytes for Cmsghdr on 64-bit Linux + 4 bytes for the fd
	oob := make([]byte, 24)

	n, oobn, recvflags, _, err := syscall.Recvmsg(int(file.Fd()), buf, oob, 0)
	if err != nil {
		return -1, nil, fmt.Errorf("recvmsg failed: %w", err)
	}

	// Check MSG_TRUNC flag to detect if data was truncated
	if recvflags&syscall.MSG_TRUNC != 0 {
		return -1, nil, fmt.Errorf("handoff data truncated (exceeded %d byte buffer); increase MaxHandoffDataSize", MaxHandoffDataSize)
	}

	// Parse control message to extract fd
	msgs, err := syscall.ParseSocketControlMessage(oob[:oobn])
	if err != nil {
		return -1, nil, fmt.Errorf("parse control message failed: %w", err)
	}

	if len(msgs) == 0 {
		return -1, nil, fmt.Errorf("no control messages received")
	}

	fds, err := syscall.ParseUnixRights(&msgs[0])
	if err != nil {
		return -1, nil, fmt.Errorf("parse unix rights failed: %w", err)
	}

	if len(fds) == 0 {
		return -1, nil, fmt.Errorf("no file descriptors received")
	}

	return fds[0], buf[:n], nil
}

// streamToClient sends an SSE response to the client.
// Replace the demo response with your LLM API integration.
func streamToClient(ctx context.Context, clientFile *os.File, handoff HandoffData) error {
	// Create net.Conn from file to enable write deadlines
	// We must write to this conn (not clientFile) for the deadline to apply
	conn, err := net.FileConn(clientFile)
	if err != nil {
		return fmt.Errorf("failed to create conn from file: %w", err)
	}
	defer conn.Close()

	// Set initial write timeout - fail fast if we can't set deadline
	if err := conn.SetWriteDeadline(time.Now().Add(WriteTimeout)); err != nil {
		return fmt.Errorf("could not set write deadline: %w", err)
	}

	writer := bufio.NewWriter(conn)

	// Send HTTP headers for SSE
	headers := []string{
		"HTTP/1.1 200 OK",
		"Content-Type: text/event-stream",
		"Cache-Control: no-cache",
		"Connection: close",
		"X-Accel-Buffering: no", // Disable buffering in nginx/proxies
	}
	for _, h := range headers {
		if _, err := fmt.Fprintf(writer, "%s\r\n", h); err != nil {
			return fmt.Errorf("failed to write header: %w", err)
		}
	}
	if _, err := fmt.Fprintf(writer, "\r\n"); err != nil {
		return fmt.Errorf("failed to write header terminator: %w", err)
	}
	if err := writer.Flush(); err != nil {
		return fmt.Errorf("failed to send headers: %w", err)
	}

	// Check for context cancellation before starting to stream
	select {
	case <-ctx.Done():
		return ctx.Err()
	default:
	}

	// Reset deadline after successful header write
	if err := conn.SetWriteDeadline(time.Now().Add(WriteTimeout)); err != nil {
		return fmt.Errorf("could not reset write deadline: %w", err)
	}

	// Stream the response
	// TODO: Replace this with your LLM API call
	return streamDemoResponse(ctx, conn, writer, handoff)
}

// streamDemoResponse simulates an LLM streaming response.
// Replace this with your actual LLM integration, e.g.:
//
//	stream, _ := openai.CreateChatCompletionStream(ctx, request)
//	for {
//	    chunk, err := stream.Recv()
//	    if err == io.EOF { break }
//	    sendSSE(conn, writer, chunk.Choices[0].Delta.Content)
//	}
func streamDemoResponse(ctx context.Context, conn net.Conn, writer *bufio.Writer, handoff HandoffData) error {
	response := fmt.Sprintf(
		"Hello! I received your prompt: %q\n\n"+
			"This response is being streamed from a daemon that received "+
			"the client socket via SCM_RIGHTS. The Apache worker has been freed.\n\n"+
			"Replace this demo with your LLM API integration.",
		truncate(handoff.Prompt, 100))

	// Stream word by word to simulate LLM token generation
	words := strings.Fields(response)
	for i, word := range words {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		// Add space before word (except first)
		content := word
		if i > 0 {
			content = " " + word
		}

		// Send SSE event (resets write deadline after each successful write)
		if err := sendSSE(conn, writer, content); err != nil {
			return err
		}

		// Simulate token generation delay
		time.Sleep(50 * time.Millisecond)
	}

	// Send completion marker
	if _, err := fmt.Fprintf(writer, "data: [DONE]\n\n"); err != nil {
		return fmt.Errorf("write failed: %w", err)
	}
	if err := writer.Flush(); err != nil {
		return fmt.Errorf("flush failed: %w", err)
	}
	return nil
}

// sendSSE sends a single SSE event with the given content.
// It resets the write deadline after each successful flush to implement
// a per-write idle timeout (not a total stream timeout).
func sendSSE(conn net.Conn, writer *bufio.Writer, content string) error {
	// json.Marshal on map[string]string cannot fail with current types;
	// keep error check as defensive programming in case payload changes.
	data, err := json.Marshal(map[string]string{"content": content})
	if err != nil {
		return fmt.Errorf("json marshal failed: %w", err)
	}
	if _, err := fmt.Fprintf(writer, "data: %s\n\n", data); err != nil {
		return fmt.Errorf("write failed: %w", err)
	}
	if err := writer.Flush(); err != nil {
		return fmt.Errorf("flush failed: %w", err)
	}
	// Reset deadline after successful write for per-write idle timeout.
	// Treat failure as fatal so we don't continue streaming without timeout protection.
	if err := conn.SetWriteDeadline(time.Now().Add(WriteTimeout)); err != nil {
		return fmt.Errorf("set write deadline failed: %w", err)
	}
	return nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
